back/app/services/surveys/results_with_logic_generator.rb

```diff
@@ -0,0 +1,223 @@
+      end
+
+      no_answer_logic_page_id = field.logic['rules']&.find { |r| r['if'] == 'no_answer' }&.dig('goto_page_id')
+      option_logic['no_answer'] = { id: "#{field.id}_no_answer", nextPageId: no_answer_logic_page_id } if no_answer_logic_page_id
+
+      option_logic.present? ? { answer: option_logic } : {}
+    end
+
+    # Replace logicNextPageId with logicNextPageNumber & add number used by FE in logic tooltip
+    # Add hidden flag to results based on logic ids supplied for filtering
+    def add_logic_to_results(results, logic_ids)
+      results_to_hide = []
+      results = results.deep_dup.map do |result|
+        field_id = result[:customFieldId]
+        if supports_page_logic? result[:inputType]
+          # Transform page logic
+          logic_next_page_id = result[:logic][:nextPageId]
+          if logic_next_page_id
+            logic_skipped_fields = logic_skipped_field_ids(results, field_id, logic_next_page_id)
+            result[:logic] = {
+              nextPageNumber: @page_numbers[logic_next_page_id],
+              numQuestionsSkipped: logic_skipped_fields.count { |f| f[:question] == true }
+            }
+            if logic_ids.include?(field_id)
+              results_to_hide += logic_skipped_fields.pluck(:id)
+            end
+          end
+        elsif supports_question_logic? result[:inputType]
+          # Transform select option logic
+          result[:logic][:answer]&.each_value do |answer|
+            logic_next_page_id = answer[:nextPageId]
+            if logic_next_page_id
+              logic_skipped_fields = logic_skipped_field_ids(results, field_id, logic_next_page_id)
+              answer[:nextPageNumber] = @page_numbers[logic_next_page_id]
+              answer[:numQuestionsSkipped] = logic_skipped_fields.count { |f| f[:question] == true }
+              answer.delete(:nextPageId)
+              if logic_ids.include?(answer[:id])
+                results_to_hide += logic_skipped_fields.pluck(:id)
+              end
+            end
+          end
+        end
+
+        result
+      end
+
+      # Now hide any results which should be hidden by the logic ids supplied for filtering
+      results.map do |result|
+        result[:hidden] = results_to_hide.include?(result[:customFieldId])
+        result
+      end
+    end
+
+    def logic_skipped_field_ids(results, field_id, goto_page_id)
+      skip = false
+      skip_from_next_page = false
+      skip_fields = []
+      results.each do |r|
+        if r[:customFieldId] == goto_page_id
+          skip = false
+          skip_from_next_page = false
+        end
+        skip = true if skip_from_next_page && r[:inputType] == 'page'
+        skip_fields << { id: r[:customFieldId], question: r[:inputType] != 'page' } if skip
+        skip_from_next_page = true if r[:customFieldId] == field_id
+      end
+      skip_fields
+    end
+
+    def change_counts_for_logic(results, survey_responses)
```

Comment:
I feel like this method is one of the hardest to understand, esp. if not familiar with this code or surveys as a whole.

I wonder if a paragraph of comments, explaining the role of the method, and a little of how it works, might be a good idea. It seems its main function is to set/change the `totalResponsesCount` and the count of `not_answered responses` (which can be lists, rather than singular, I think).